Clean up overrides.ts

Philipp91 opened the pull request:
- Use inline imports to make the file less verbose and easier to change in the future.
- Sort alphabetically instead of arbitrarily.
- Name styles consistently like `MuiPickers<component_name>`, except for `MuiPickersBasePicker`.

Philipp91 committed:
Clean up overrides.ts

- Use inline imports to make the file less verbose and easier to change in the future.
- Sort alphabetically instead of arbitrarily.
- Name styles consistently like `Mui<component_name>`, except for `MuiBasePicker`.

dmtrKovalenko commented on lib/src/DatePicker/DatePickerToolbar.tsx:
```diff
@@ -18,7 +18,7 @@ export const useStyles = makeStyles(
     },
   },
-  { name: 'MuiPickersDatePickerRoot' }
+  { name: 'MuiPickersDatePickerToolbar' }
```
Maybe we need some reusable function that will build a standardized name. Something like
```ts
const buildStyleName = (component: React.ComponentType<any>) => {
   return `MuiPickers${component.displayName || component.name}`
}
```

@oliviertassinari what do your think?

Philipp91 replied:
How about making the second argument to `makeStyles()` a component?

oliviertassinari replied:
I think that we should drop the "Pickers" prefix to match the exported name. What would be wrong with?

```suggestion
  { name: 'MuiDatePickerToolbar' }
```

Actually, this would be even better as the component is private:

```suggestion
  { name: 'PrivateDatePickerToolbar' }
```

Philipp91 replied:
What do you mean by "even better"?
#1707 makes this public

Philipp91 replied:
Also regarding the `buildStyleName()` function, especially if the prefix is now just `Mui` and nost `MuiPickers`, I think this library should just do the same as the main material-ui library. Does that one have such function?

oliviertassinari replied:
> #1707 makes this public

I think that the change should be atomic. If we make a new component public, it should be done all at once (same pull request}. This would avoid the cases were we forget/revert/etc.

oliviertassinari replied:
> Does that one have such function?

Great point, we don't. I think that a string duplication is good enough.

oliviertassinari commented on lib/src/DateRangePicker/DateRangePickerToolbar.tsx:
```diff
@@ -17,7 +17,7 @@ export const useStyles = makeStyles(
     },
   },
-  { name: 'MuiPickersDatePickerRoot' }
+  { name: 'MuiPickersDateRangePickerToolbar' }
```
```suggestion
  { name: 'MuiDateRangePickerToolbar' }
```

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerTabs.tsx:
```diff
@@ -44,7 +44,7 @@ export const useStyles = makeStyles(
     };
   },
-  { name: 'MuiPickerDTTabs' }
+  { name: 'MuiPickersDateTimePickerTabs' }
```
```suggestion
  { name: 'PrivateDateTimePickerTabs' }
```

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -34,7 +34,7 @@ export const useStyles = makeStyles(
     },
   }),
-  { name: 'MuiPickerDTToolbar' }
+  { name: 'MuiPickersDateTimePickerToolbar' }
```
```suggestion
  { name: 'MuiDateTimePickerToolbar' }
```

Philipp91 replied:
If the prefix is really just `Mui` and not `MuiPickers`, that might cause clashes on things like the `ModalDialog`. Currently, that one is not exported, but I'm not sure if even `PrivateModalDialog` could clash with another material-ui-based library that has its own ModalDialog?

Same for `Toolbar`, `ToolbarButton` or `StaticWrapper`.

oliviertassinari commented on lib/src/typings/overrides.ts:
```diff
@@ -39,34 +17,29 @@ type Classes<T> = Partial<
-  MuiPickersDateRangePickerInput?: Classes<
-    typeof import('../DateRangePicker/DateRangePickerInput').useStyles
-  >;
+  MuiPickersBasePickerStyles?: Classes<typeof import ('../Picker/Picker').useStyles>;
```
Most of the exported override keys here seem to be from private components, so to be removed.
Also, in https://github.com/mui-org/material-ui/issues/19427, it seems that we make the DX simpler for TypeScript users.

Philipp91 replied:
So the idea there is to stop having this central "registry" in overrides.ts and instead only declaring the export in the respective component's file?
(Something like that would make it easier to keep privateness in sync.)

Philipp91 replied:
What makes a component public (or private)? For instance, if I don't see the `ArrowSwitcher` component exported from `src/index.ts`, does that mean that it's private? And also, assuming that was right, should I remove it from `overrides.ts` and call its styles `PrivateArrowSwitcher` instead?

Will we also have a `buildStyleName()` variant or parameter for private components?

oliviertassinari replied:
> What makes a component public (or private)?

Great question. So far on https://github.com/mui-org/material-ui it has been as simple as: Is the module exported from the barrel index? If exported public, if not, private.

Philipp91 replied:
Side question: Is it intended that things like `defaultReduceAnimations` are exported from the top-level `index.ts`?

Philipp91 replied:
Do the `overrides.ts` exports and the `index.ts` exports necessarily have to be in sync? Is that a best practice? What are the advantages?

Anyway I pushed a commit that hopefully does that.

oliviertassinari replied:
So far, we have been keeping them in sync. So unless there is a good reason to deviate from this rule, I would be in favor of sticking to it.

Philipp91 committed:
Remove private components from overrides.ts

Also add one (ClockView) that is public and ensure that all the styles names are either prefixed by `Mui` for public components or by `Private` for private components.

Philipp91 commented:
There are a number of questions that haven't been answered:
> Side question: Is it intended that things like defaultReduceAnimations are exported from the top-level index.ts?
> If the prefix is really just Mui and not MuiPickers, that might cause clashes on things like the ModalDialog.
> Maybe we need some reusable function that will build a standardized name. Something like ... `buildStyleName` ...

Though it's quite possible that you think it's fine to leave as is. Otherwise please let me know if you want me to change something in this PR.

oliviertassinari commented:
> Side question: Is it intended that things like defaultReduceAnimations are exported from the top-level index.ts?

@Philipp91 I think that it was exposed by chance.

> If the prefix is really just Mui and not MuiPickers, that might cause clashes on things like the ModalDialog.

The picker components are part of the official set of Material-UI components. I think that the prefix should be `Mui`.

> Maybe we need some reusable function that will build a standardized name. Something like ... buildStyleName ...

What's wrong with a duplication of the name of the component?

Philipp91 commented:
Okay, then this PR should be fine as is.

dmtrKovalenko commented:
@oliviertassinari we cannot ship this PR. 
We are not providing any other styling solution different from global overrides. So we **must** make all of our components style-able by global overrides.

I am so sorry for being too late @Philipp91. You are doing awesome contribution, thank you!

dmtrKovalenko commented:
Even if we will provide `classes` prop. We need to make all the components like `Calendar` accept global overrides, just because it is useful as feature

oliviertassinari commented:
> So we must make all of our components style-able by global overrides.

@dmtrKovalenko The concern is interesting. The way we have solved that problem with the other components was by either 1. exposing the lower-level building blocks, allowing people to compose them or 2. to keep all the logic under a single public style sheet and hide a bit into private style sheets with the option to provide custom implementations.
However, this component is different. It's the first time I encounter a component with this set of challenges.

Any objection to make all the components that expose their styles public?

I have noticed a couple of styles that are private, is it expected?

<img width="368" alt="Capture d’écran 2020-05-08 à 00 32 22" src="https://user-images.githubusercontent.com/3165635/81350962-7427a880-90c3-11ea-829d-70776c58770b.png">

dmtrKovalenko commented:
1. I don’t really know about all the components. But it looks useful from the custom design system perspective. If we can design anything in 1 place it is easier to reuse custom theme. 

2. Private styles are unexpected.

oliviertassinari commented:
How can we move forward? Maybe we should be less ambitious with the changes here. @Philipp91 Was was the original issue you were going after?

Philipp91 commented:
My ultimate goal was exporting the toolbar components (#1707). It doesn't really matter (for my use case) whether their styles are exported too or not.

oliviertassinari commented:
Ok cool, in this case, maybe we should close and leave this pull request to the core team?

Philipp91 commented:
Sure, I'll let you take it over.
